### PR TITLE
Add ability to choose merge seleting algorithm

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduced MergeTree setting which allow to change merge selecting algorithm. However we still have only one algorithm and it's mostly for future experiments.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit f6ca1613d85dd5755affeef0064d72f89e4b77fc with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70647/f6ca1613d85dd5755affeef0064d72f89e4b77fc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ad7c74c549343391b5320bcf95dc2b44b4f04d40/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeSelectors/MergeSelectorFactory.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    using EnumToName = std::unordered_map<MergeSelectorAlgorithm, std::string>;
+
+    CreatorByNameMap creators;
+    EnumToName enum_to_name_mapping;
+    MergeSelectorFactory() = default;
+public:
+    static MergeSelectorFactory & instance();
+
+    MergeSelectorPtr get(const std::string & name) const;
+    MergeSelectorPtr get(const std::string & name, const std::any & settings) const;
```
Comment (Michicosun) [inline src/Storages/MergeTree/MergeSelectors/MergeSelectorFactory.h:33]: maybe merge into single method with defaulted parameter?

**Review by Michicosun — APPROVED** [review]
LGTM

File: src/Storages/MergeTree/MergeSelectors/MergeSelectorFactory.h (hunk 1 above)
Comment (alesapin) [reply src/Storages/MergeTree/MergeSelectors/MergeSelectorFactory.h:33]: yep, 100%

**Comment by alesapin** [comment]
Waiting for builds only

**Pull request merged**